src/lua/datetime.lua

```diff
@@ -0,0 +1,623 @@
+        sec_int, fraction = math_modf(ts)
+        -- if there is one of nsec, usec, msec provided
+        -- then ignore fraction in timestamp
+        -- otherwise - use nsec, usec, or msec
+        if count_usec == 0 then
+            nsec = fraction * 1e9
+        end
+
+        self.secs = sec_int
+        self.nsec = nsec
+
+        return self
+    end
+
+    local offset0 = self.tzoffset
+    local offset = obj.tzoffset
+    if offset ~= nil then
+        offset = get_timezone(offset)
+        check_range(offset, {-720, 720}, 'tzoffset')
+        -- self.tzoffset = offset
+    end
+
+    if obj.tz ~= nil then
+        nyi('tz')
+    end
+
+    -- .year, .month, .day
+    if ymd then
+        datetime_ymd_update(self, y or y0, M or M0, d or d0, offset or offset0)
+    end
+
+    -- .hour, .minute, .second
+    if hms then
+        datetime_hms_update(self, h or h0, m or m0, sec or sec0, offset or offset0)
+    end
+
+    self.epoch, self.nsec = normalize_nsec(self.epoch, self.nsec)
+
+    if offset ~= nil then
+        self.tzoffset = offset
+    end
+
+    return self
+end
+
+local strfmt_sz = 128
+local strfmt_buff = ffi.new('char[?]', strfmt_sz)
+
+local function datetime_strftime(self, fmt)
+    check_date(self, "datetime.strftime()")
+    builtin.datetime_strftime(strfmt_buff, strfmt_sz, fmt, self)
+    return ffi.string(strfmt_buff)
+end
+
+local function datetime_format(self, fmt)
+    if fmt ~= nil then
+        return datetime_strftime(self, fmt)
+    else
+        return datetime_tostring(self)
+    end
+end
+
+ffi.metatype(datetime_t, {
+    __tostring = datetime_tostring,
+    __serialize = datetime_serialize,
+    __eq = datetime_eq,
+    __lt = datetime_lt,
+    __le = datetime_le,
+    __index = {
+        epoch = function(self) return self.epoch end,
+        timestamp = function(self) return self.epoch + self.nsec / 1e9 end,
+
+        nsec = function(self) return self.nsec end,
+        usec = function(self) return self.nsec / 1e3 end,
+        msec = function(self) return self.nsec / 1e6 end,
+
+        dt = function(self) return local_dt(self) end,
+        year = function(self) return builtin.tnt_dt_year(local_dt(self)) end,
+        month = function(self) return builtin.tnt_dt_month(local_dt(self)) end,
+
+        yday = function(self) return builtin.tnt_dt_doy(local_dt(self)) end,
+        wday = function(self)
+            return ffi.cast('int32_t', builtin.tnt_dt_dow(local_dt(self)))
+        end,
+        day = function(self)
+            return builtin.tnt_dt_dom(local_dt(self))
+        end,
+        hour = function(self) return math_floor((local_secs(self) / 3600) % 24) end,
+        minute = function(self) return math_floor((local_secs(self) / 60) % 60) end,
+        second = function(self) return self.epoch % 60 end,
+
+        format = datetime_format,
+        totable = datetime_totable,
+        set = datetime_set,
+    }
+})
+
+return setmetatable(
+    {
+        new         = datetime_new,
+
```

Comment:
Please, drop the excess empty lines.

Reply:
These are not excess lines, they are separators for function groups. [Please, see how they look like in the final form at the end of branch] I'd prefer to keep them visually separated.

Reply:
On top of the branch I see 3 functions, all with a lot of excess spaces before `=` and with excess lines before and after them. Please, drop it all. The empty lines can be added later when (and if) necessary, it won't change the existing lines anyway.